// application/Initializer.ts
import { dbConfig } from "config/AppConfig";
const _wait = require("wait.for");
const WaitFor = require("framework/WaitFor");
const Sequelize = require('sequelize');
const Promise = require("promise");
import logger from "framework/Logger";;

export const sequelize = new Sequelize(dbConfig.database, dbConfig.username, dbConfig.password, {
    host: dbConfig.host,
    dialect: 'postgres',
    port: dbConfig.port,
    pool: {
        max: 100,
        min: 0,
        idle: 10* 1000,
        evict: 60 * 1000
    },
});


export default function Initializer() {
    return new Promise(function(resolve: any, reject: any) {
        const interval = setInterval(() => {
            sequelize.authenticate()
                .then(function(error) {
                    if (!error) {
                        clearInterval(interval);
                        resolve("CONNECTED TO DB");
                    } else {
                        console.log("PING DATABASES");
                    }
                })
        }, 1000 * 5);
    });
}
