// framework/DateUtil.ts
export function MakeDateUtil() {
    return new DateUtil();
}

export class DateUtil {
    private date: Date;

    public setTimeStamp(timestamp: number) {
        this.date = new Date(Number(timestamp));
        return this;
    }

    public setDate(date: Date) {
        this.date = date;
        return this;
    }

    public setDateFromStringDate(date: string) {
        this.date = new Date(date);
        return this;
    }

    public getHours() {
        return this.date.getHours();
    }

    public getMinutes() {
        return this.date.getMinutes();
    }

    public getSeconds() {
        return this.date.getSeconds();
    }

    public getDate() {
        return this.date.getDate();
    }

    public getMonth() {
        return this.date.getMonth();
    }

    public getYear() {
        return this.date.getFullYear();
    }

    public plusMoreMinutes(minutes: number) {
        return new Date(new Date().getTime() + minutes * 60 * 1000);
    }

    public getTimes() {
        return this.date.getTime();
    }
}
